drivers/gpio/gpio_emul.c

```diff
@@ -0,0 +1,596 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_gpio_emul
+
+#include <device.h>
+#include <drivers/gpio.h>
+#include <drivers/gpio/gpio_emul.h>
+#include <errno.h>
+#include <zephyr.h>
+
+#include "gpio_utils.h"
+
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_emul);
+
+#define GPIO_EMUL_INT_BITMASK						\
+	(GPIO_INT_DISABLE | GPIO_INT_ENABLE | GPIO_INT_LEVELS_LOGICAL |	\
+	 GPIO_INT_EDGE | GPIO_INT_LOW_0 | GPIO_INT_HIGH_1)
+
+/**
+ * @brief Emulated GPIO controller configuration data
+ *
+ * This structure contains all of the state for a given emulated GPIO
+ * controller as well as all of the pins associated with it.
+ *
+ * The @a flags member is a pointer to an array which is @a num_pins in size.
+ *
+ * @a num_pins must be in the range [1, @ref GPIO_MAX_PINS_PER_PORT].
+ *
+ * Pin direction as well as other pin properties are set using
+ * specific bits in @a flags. For more details, see @ref gpio_interface.
+ *
+ * Changes are synchronized using @ref gpio_emul_data.mu.
+ */
+struct gpio_emul_config {
+	/** Common @ref gpio_driver_config */
+	const struct gpio_driver_config common;
+	/** Number of pins available in the given GPIO controller instance */
+	const gpio_pin_t num_pins;
+};
+
+/**
+ * @brief Emulated GPIO controller data
+ *
+ * This structure contains data structures used by a emulated GPIO
+ * controller.
+ *
+ * If the application wishes to specify a "wiring" for the emulated
+ * GPIO, then a @ref gpio_callback_handler_t should be registered using
+ * @ref gpio_add_callback.
+ *
+ * Changes are to @ref gpio_emul_data and @ref gpio_emul_config are
+ * synchronized using @a mu.
+ */
+struct gpio_emul_data {
+	/** Common @ref gpio_driver_data */
+	struct gpio_driver_data common;
+	/** Pointer to an array of flags is @a num_pins in size */
+	uint32_t *flags;
+	/** Input values for each pin */
+	gpio_port_value_t input_vals;
+	/** Output values for each pin */
+	gpio_port_value_t output_vals;
+	/** Interrupt status for each pin */
+	gpio_port_pins_t interrupts;
+	/** Mutex to synchronize accesses to driver data and config */
+	struct k_mutex mu;
+	/** Singly-linked list of callbacks associated with the controller */
+	sys_slist_t callbacks;
+};
+
+/**
+ * @brief Obtain a mask of pins that match all of the provided @p flags
+ *
+ * Use this function to see which pins match the current GPIO configuration.
+ *
+ * The caller must ensure that @ref gpio_emul_data.mu is locked.
+ *
+ * @param port The emulated GPIO device pointer
+ * @param mask A mask of flags to match
+ * @param flags The flags to match
+ *
+ * @return a mask of the pins with matching @p flags
+ */
+static gpio_port_pins_t
+get_pins_with_flags(const struct device *port, uint32_t mask,
+	uint32_t flags)
+{
+	size_t i;
+	gpio_port_pins_t matched;
```

Comment:
I'd rather see `matched` initialized here instead of the entry to the for loop.  It just looks weird there.